File: walking_robot_sim/src/quadropted_controller_cpp/src/controllers/rest_controller.cpp
#include "quadropted_controller_cpp/controllers/rest_controller.hpp"

#include <cmath>

#include "quadropted_controller_cpp/utils/rotation_matrices.hpp"

namespace quadropted {

RestController::RestController(Eigen::MatrixXd default_stance)
    : default_stance_(std::move(default_stance)), pid_(0.75, 2.29, 0.0), use_imu_(false), pid_last_time_(0.0) {}

void RestController::reset() {
    pid_.reset(0.0);
    pid_last_time_ = 0.0;
}

Eigen::MatrixXd RestController::step(const State& state, const Command& cmd) {
    Eigen::MatrixXd temp = default_stance_;
    temp.row(2).setConstant(cmd.robot_height);

    if (use_imu_) {
        // Timestamp не критичен для REST — PID используется только для integral
        auto compensation = pid_.run(state.imu_roll, state.imu_pitch, pid_last_time_);
        pid_last_time_ += 0.02;  // фиксированный шаг как в Python
        double roll_compensation = -compensation[0];
        double pitch_compensation = -compensation[1];
        Eigen::Matrix3d rot = rotxyz(roll_compensation, pitch_compensation, 0.0);
        temp = rot * temp;
    }

    return temp;
}

}  // namespace quadropted

File: walking_robot_sim/src/tests/old/RobotController/RestController.py
#!/usr/bin/env python3
# Author: lnotspotl, abutalipovvv
import rclpy
import numpy as np
from RoboticsUtilities.Transformations import rotxyz
from .PIDController import PID_controller

class RestController:
    def __init__(self, default_stance):
        self.def_stance = default_stance

        # TODO: configure kp, ki и kd
        #                                     kp     ki    kd
        self.pid_controller = PID_controller(0.75, 2.29, 0.0)
        self.use_imu = False
        self.use_button = True
        self.pid_controller.reset()

    def updateStateCommand(self, msg, state, command):
        
        state.body_local_position[0] = msg.axes[2] * 0.08
        state.body_local_position[1] = msg.axes[5] * 0.08
        state.body_local_position[2] = msg.axes[1] * 0.08

        
        state.body_local_orientation[0] = msg.axes[0] * 0.8
        state.body_local_orientation[1] = msg.axes[4] * 0.6
        state.body_local_orientation[2] = msg.axes[3] * 0.6

        if self.use_button:
            if msg.buttons[7]:
                self.use_imu = not self.use_imu
                self.use_button = False
                print(f"Rest Controller - Использование компенсации крена/тангажа: {self.use_imu}")

        if not self.use_button:
            if not msg.buttons[7]:
                self.use_button = True

    @property
    def default_stance(self):
        return self.def_stance

    def step(self, state, command):
        temp = self.default_stance
        temp[2] = [command.robot_height] * 4


        if self.use_imu:
            compensation = self.pid_controller.run(state.imu_roll, state.imu_pitch)
            roll_compensation = -compensation[0]
            pitch_compensation = -compensation[1]

            rot = rotxyz(roll_compensation, pitch_compensation, 0)
            temp = np.matmul(rot, temp)

        return temp

    def run(self, state, command):
        state.foot_locations = self.step(state, command)
        return state.foot_locations
